### ISSUE TITLE
How to improve `surface_integral` kernels

### ISSUE THREAD

**Issue by huiyuxie** [trixi-gpu/TrixiCUDA.jl#5]
This is not a typical issue as no errors ever occurred. But I think it is worth to bring up this issue here.
First, I will give some background related to this issue. See example
```Julia
using CUDA

function foo!(a, b)
    i = (blockIdx().x - 1) * blockDim().x + threadIdx().x

    if (i <= 4)
        @inbounds begin
            for ii in 1:4
                a[ii] += b[ii]
            end
        end
    end

    return nothing
end

a = CUDA.ones(4)
b = CUDA.ones(4)
@cuda threads = 4 foo!(a, b)
```
And the result is given as 
```Julia
julia> a
4-element CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
 2.0
 2.0
 2.0
 2.0
```
I used to think that parallel computing is kind of something thread-safe, that is, in this example, each value of `a` should be `5.0` instead of `2.0`. As I thought even though the computation is parallel, it should immediately write any data changes to memory based on the order of completion after each thread completes its unit of computation, but the fact is not. So here some problems came for the `surface integral` kernels.
The CPU code for calculating `surface integral` in 2D and 3D are 
https://github.com/huiyuxie/Trixi.jl/blob/fdccbb17aaddd31a935ee7560fd4ea506a6d93ad/src/solvers/dgsem_tree/dg_2d.jl#L1076-L1115
https://github.com/huiyuxie/Trixi.jl/blob/fdccbb17aaddd31a935ee7560fd4ea506a6d93ad/src/solvers/dgsem_tree/dg_3d.jl#L1299-L1349
So if we directly write something like below will cause some problems 
https://github.com/huiyuxie/trixi_cuda/blob/f1adcdf47d5840f17bed640f08af48bcfe0755c9/cuda_dg_2d.jl#L265-L281
https://github.com/huiyuxie/trixi_cuda/blob/f1adcdf47d5840f17bed640f08af48bcfe0755c9/cuda_dg_3d.jl#L278-L299
since the same entry from `du` will be computed in parallel among different threads, like `j=size(du, 2), k=1` and `j=1, k=1` in 2D (i.e. `du[.., 1, size(du,2), 1]` has been computed repeatedly) , and like `j1=1, j2=size(du, 2), k=1` and `j1=1, j2=1, k=1` in 3D (i.e. `du[.., 1, 1, size(du,2), 1]` has been computed repeatedly). As mentioned before, though these entries have been computed repeatedly but not all the data change will be write to the memory and thus cause data loss. 
Thus, currently my solutions are like 
https://github.com/huiyuxie/trixi_cuda/blob/f1adcdf47d5840f17bed640f08af48bcfe0755c9/cuda_dg_2d.jl#L283-L313
https://github.com/huiyuxie/trixi_cuda/blob/f1adcdf47d5840f17bed640f08af48bcfe0755c9/cuda_dg_3d.jl#L301-L356
separating the target entries in different kernels and then run kernels in series. 
But I really don't like this way as it 1) definitely increases run time, and 2) (seems like) could be done in single kernel but now not, and 3) makes code ugly. So is there any other way to improve this kind of kernel? Now I have no better ideas so this issue is on hold.

**Comment by ranocha** [comment]
Do you have some benchmarks of these?

It may be worth trying to rewrite the implementations from a scatter-type approach (choosing the indices of the sources on the right-hand side) to a gather-type approach where you choose the indices of the destinations on the left-hand side.

**Comment by huiyuxie** [comment]
Thanks for your advice! New kernels for `surface integral` are like 
https://github.com/huiyuxie/trixi_cuda/blob/beefbce72286c01beb216d00e3a467eb7eea5c32/cuda_dg_1d.jl#L353-L367
https://github.com/huiyuxie/trixi_cuda/blob/beefbce72286c01beb216d00e3a467eb7eea5c32/cuda_dg_2d.jl#L422-L443
https://github.com/huiyuxie/trixi_cuda/blob/beefbce72286c01beb216d00e3a467eb7eea5c32/cuda_dg_3d.jl#L470-L496
and benchmarks are done for `test/hypdiff_harmonic_nonperiodic_1d.jl`, `test/hypdiff_harmonic_nonperiodic_2d.jl`, and `test/hypdiff_nonperiodic_3d.jl` and results are like 
```Julia
# 1D original 
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  10.441 μs … 79.272 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     11.776 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   12.238 μs ±  2.241 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%
 
 # 1D new 
 BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  10.978 μs … 80.150 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     11.913 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   12.296 μs ±  1.646 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%
```
```Julia
# 2D original
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  21.437 μs …  11.447 ms  ┊ GC (min … max): 0.00% … 99.25%
 Time  (median):     24.162 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   26.002 μs ± 114.250 μs  ┊ GC (mean ± σ):  4.37% ±  0.99%
 
 # 2D new
 BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  12.043 μs … 89.029 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     13.349 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   13.848 μs ±  2.041 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%
```
```Julia
# 3D original
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  36.816 μs …  11.230 ms  ┊ GC (min … max): 0.00% … 98.82%
 Time  (median):     40.016 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   42.038 μs ± 111.941 μs  ┊ GC (mean ± σ):  2.64% ±  0.99%
 
 # 3D new
 BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  12.826 μs … 81.850 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     14.095 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   14.615 μs ±  2.003 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%
```
For the 1D case, the new implementation is slower as we set up a larger kernel compared to the original one. But for the 2D and 3D cases, the new implementation is nearly half and a third as fast as the original one respectively as we set up fewer kernels. I would like to apply the new version of the kernels to 1D, 2D, and 3D (though it is slightly slower for 1D) since I want to maintain code consistency.

**Comment by huiyuxie** [comment]
Sorry! I think I should have continued to consider this issue, as it is not very hard. But I have already spent long time trying to understand why there are persistent calculation errors. So I think it would be better to drop this issue here and focus on implementing other kernels as the improvement should be an issue that can be addressed later. If you need more benchmark results (like testing on other equations) please let me know. Thank you so much for your help!

**Comment by ranocha** [comment]
No, that's fine with me :+1:
